wagtail/core/models.py

```diff
@@ -3185,6 +3471,178 @@ def get_comment(self):
+                    'title': self.workflow_state.workflow.name,
+                    'status': self.status,
+                    'task': {
+                        'id': self.task.id,
+                        'title': self.task.name,
+                    },
+                    'next': next_task_data,
+                }
+            },
+            revision=self.page_revision
+        )
+
     class Meta:
         verbose_name = _('Task state')
         verbose_name_plural = _('Task states')
+
+
+class LogEntryManager(models.Manager):
+
+    def log_action(self, instance, action, **kwargs):
+        """
+        :param instance: The model instance we are logging an action for
+        :param action: The action. Should be namespaced to app (e.g. wagtail.create, wagtail.workflow.start)
+        :param kwargs: Addition fields to for the LogEntry model
+            - user: The user performing the action
+            - title: the instance title
+            - data:
+            - revision: a PageRevision instance, if the instance is a
+            - created, published, unpublished, content_changed, deleted - Boolean flags
+        :return: The new log entry
+        """
+        data = kwargs.pop('data', '')
+        title = kwargs.pop('title', None)
+        if not title:
+            if hasattr(instance, 'get_admin_display_title'):
+                title = instance.get_admin_display_title()
+            else:
+                title = str(instance)
+        return self.model.objects.create(
+            content_type=ContentType.objects.get_for_model(instance, for_concrete_model=False),
+            object_id=instance.pk,
+            object_title=title,
+            action=action,
+            timestamp=timezone.now(),
+            data_json=json.dumps(data),
+            **kwargs,
+        )
+
+    def get_for_model(self, model):
+        # Return empty queryset if the given object is not valid.
+        if not issubclass(model, models.Model):
+            return self.none()
+
+        ct = ContentType.objects.get_for_model(model)
+
+        return self.filter(content_type=ct)
+
+    def get_for_instance(self, instance):
+        ct = ContentType.objects.get_for_model(instance, for_concrete_model=False)
+        return self.filter(content_type=ct, object_id=instance.pk)
+
+    def get_for_user(self, user_id):
+        return self.filter(user=user_id)
+
+    def get_pages(self):
+        content_types = ContentType.objects.get_for_models(*get_page_models()).values()
+        return self.filter(content_type__in=content_types)
+
+    def get_pages_for_user(self, user_id):
+        return self.get_pages().filter(user=user_id)
+
+
+class LogEntry(models.Model):
+    content_type = models.ForeignKey(
+        ContentType,
+        models.SET_NULL,
+        verbose_name=_('content type'),
+        blank=True, null=True,
+        related_name='+',
+    )
+    object_id = models.CharField(max_length=255, blank=False, db_index=True)
```

Comment:
I'm thinking this generic foreign key is probably more trouble than it's worth:

* There are lots of places (e.g. `LogEntriesView`) where we're filtering on `object_id` but don't have a corresponding filter on `content_type`, which need fixing if we're ever going to have log entries for non-page types
* It gets in the way of efficient querying - e.g. `LogEntriesView` has to retrieve the entire `explorable_pages()` dataset (which may contain many thousands of IDs) to convert to strings to pass back to the next query. With a non-generic foreign key, Django would (hopefully) be able to compile this as a subquery that the database can optimise.
* Many of the definitions here are page-specific - realistically, if we extended logging to non-page models, isn't it more likely that they'd use a separate model (possibly with a common abstract base model / mixin)?

Reply:
I have debated this for a bit. Decided to make it generic to accomodate non-integer PKs.
FWIW, `LogEntriesView` does `LogEntry.objects.get_pages()` which limits the qs to `Page`-derived content types

> It gets in the way of efficient querying
Great shout on the `explorable_pages()` dataset.

> Many of the definitions here are page-specific - realistically, if we extended logging to non-page models, isn't it more likely that they'd use a separate model (possibly with a common abstract base model / mixin)?

I see where you are going with this. In hindsight a common abstract model with non-page model specific implementations is probably more elegant. So `LogEntry` becomes `BaseLogEntry`, with a `PageLogEntry` models with a page FK with `on_delete=models.DO_NOTHING`.

Can work with that

Reply:
I guess the downside of moving to an abstract model / mixin approach is that pulling it all together in "site history" requires jumping through hoops. 

Do we tackle that when we introduce logging for non-Page model? e.g. by adding a separate filter that picks up the various instances (page, image, doc etc) and provide the appropriate querysets, defaulting to `PageLogEntry` objects?

Reply:
Went with abstract base class.